octodns/provider/googlecloud.py

```diff
@@ -0,0 +1,343 @@
+            elif class_name == 'Update':
+                gcloud_changes.delete_record_set(
+                    _rrset_func(gcloud_zone, change.existing))
+                gcloud_changes.add_record_set(
+                    _rrset_func(gcloud_zone, change.new))
+            else:
+                raise RuntimeError('Change type "{}" for change "{!s}" '
+                                   'is none of "Create", "Delete" or "Update'
+                                   .format(class_name, change))
+
+        gcloud_changes.create()
+        i = 1
+        while gcloud_changes.status != 'done':
+            self.log.debug("Waiting for changes to complete")
+            time.sleep(i)
+            gcloud_changes.reload()
+            if i < 30:
+                i += 2
+
+    def _create_gcloud_zone(self, dns_name):
+        """Creates a google cloud ManagedZone with dns_name, and zone named
+            derived from it. calls .create() method and returns it.
+
+            :param dns_name: fqdn of zone to create
+            :type  dns_name: str
+
+            :type return: new google.cloud.dns.ManagedZone
+        """
+        # Zone name must begin with a letter, end with a letter or digit,
+        # and only contain lowercase letters, digits or dashes
+        zone_name = re.sub("[^a-z0-9-]", "",
+                           dns_name[:-1].replace('.', "-"))
+        # make sure that the end result did not end up wo leading letter
+        if re.match('[^a-z]', zone_name[0]):
+            # I cannot think of a situation where a zone name derived from
+            # a domain name would'nt start with leading letter and thereby
+            # violate the constraint, however if such a situation is
+            # encountered, add a leading "a" here.
+            zone_name = "a%s" % zone_name
+
+        gcloud_zone = self.gcloud_client.zone(
+            name=zone_name,
+            dns_name=dns_name
+        )
+        gcloud_zone.create(client=self.gcloud_client)
+
+        self.log.info("Created zone %s. Fqdn %s." %
+                      (zone_name, dns_name))
+
+        return gcloud_zone
+
+    def _get_gcloud_records(self, gcloud_zone, page_token=None):
+        """ Generator function which yields ResourceRecordSet for the managed
+            gcloud zone, until there are no more records to pull.
+
+            :param gcloud_zone: zone to pull records from
+            :type gcloud_zone: google.cloud.dns.ManagedZone
+            :param page_token: page token for the page to get
+
+            :return: a resource record set
+            :type return: google.cloud.dns.ResourceRecordSet
+        """
+        gcloud_iterator = gcloud_zone.list_resource_record_sets(
+            page_token=page_token)
+        for gcloud_record in gcloud_iterator:
+            yield gcloud_record
+        # This is to get results which may be on a "paged" page.
+        # (if more than max_results) entries.
+        if gcloud_iterator.next_page_token:
```

Comment:
If I'm reading this correctly I believe this may only handle the case when there's 2 pages of results. Rough stab at what may help here if that's correct:

```python
while True:
    gcloud_iterator = gcloud_zone.list_resource_record_sets(page_token=page_token)
    for gcloud_record in gcloud_iterator:
        yield gcloud_record
    page_token = gcloud_iterator.next_page_token
    # This assumes that it's None when there's no more pages
    if page_token is None: 
        break
```

Reply:
Aactually this function will simply call itself recursively until the iterator returned does not have the "next_page_token". I did update the unit test to fetch a result which is on a simulated "page 3" and it does work.

Reply:
Hrm. I think I was confused about the yields. I guess that would involve recursively yielding as well? That's not something I've run across before. I don't know enough about the implementation details of python's `yield` to know if that's a problem, but it seems like it'd at least be sub-optimal. 

I feel like the `while True` walk pages and yield the records is a bit simpler and easier to understand. Thoughts?

Reply:
Yes. It will recursively yield until there are no more results. With python 3.3 comes the `yield from` expression "(allowing a generator to delegate part of its operations to another generator)"  which I think has been working great for making yields from recursive functions. this is python27 though and it will yield the results it got by recursively calling itself. Anyway, in this scenario I don't think it's going to have any measurable performance differences compared to a `while True` walk solution proposed by you, and I don't know which is easier to read, but that can be because I know how the function I wrote works and may be biased. Also I have no strong opinions in favour of either solution, only that the one I have written is written already, and thus requires no more work on my end for now.